### PR TITLE
SNOW-1060154: Add telemetry for query plan height and num duplicate subtrees

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-1060154

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [x] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR add `plan_height` and `num_duplicate_nodes` in `SnowflakePlan` and `Selectable` class to compute characteristics of the query plan tree and report these in telemetry

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:671]: linter fix

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:329]: `find_duplicate_subtrees` looks expensive. This seems to run every time a `SnowflakePlan` is created?

**Comment by sfc-gh-jdu** [comment]
I assume it will be rebased after #1295 being merged?

**Comment by sfc-gh-aalam** [comment]
> I assume it will be rebased after #1295 being merged?

yes

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:329]: this is called when evaluating `df_collect_api_telemetry` so only when action statement is called.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:217]: I see. I know we also added `children` in `SetStatement` and `SelectStatement` in PR #1295 to help us traverse the plan with better accuracy, for same reason we add it here. I'm not sure if we can rely `source_plan` to cover all cases.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:329]: We can do a performance tests for the client side.

File: src/snowflake/snowpark/_internal/telemetry.py
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/telemetry.py]: done

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:217]: actually I just realized that we can't just add children for them, because when we resolve a logic plan here https://github.com/snowflakedb/snowpark-python/blob/8a6b98ad662010670f8e4b750fe791c4ba799232/src/snowflake/snowpark/_internal/analyzer/analyzer.py#L727, we always iterate through it's children. So if a SnowflakePlan or Selectable object is passed into this function (I verified Selectable will be passed into it), we will resolve them again, which are unnecessary (though it seems no negative effect). So it'd better to use parameter to protect this change.

File: src/snowflake/snowpark/_internal/telemetry.py
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/telemetry.py:159]: Please do a performance test for the client side.

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: A question is how do we query this telemetry?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: Query below can be used to extract current sql simplifier telemetry. Since we are reporting query plan height and num duplicate nodes within the same telemetry json, both these values can be extracted similarly. 
```
select d, seq, value from (select parse_json(data):api_calls as d from snowscience.snowpark.client_telemetry_snowpark_function where created_on > dateadd(days, -5, current_timestamp) limit 100)  t1, lateral flatten(input => t1.d)
where value like '%sql_simplifier_enabled%';
```

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: > Please do a performance test for the client side.

Let me do a test and report back

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: I did a test run with dummy ctc-repro with 50 repeated self joins the calculation time was about 0.2 seconds for the largest graph
```
sending telemetry: {'name': 'Session.sql', 'sql_simplifier_enabled': True, 'query_plan_height': 2, 'query_plan_num_duplicate_nodes': 0}, elapsed time = 0.0002963542938232422
sending telemetry: {'name': 'Session.table', 'sql_simplifier_enabled': True, 'query_plan_height': 26, 'query_plan_num_duplicate_nodes': 21}, elapsed time = 0.2114887237548828
sending telemetry: {'name': 'Session.sql', 'sql_simplifier_enabled': True, 'query_plan_height': 4, 'query_plan_num_duplicate_nodes': 0}, elapsed time = 0.0007040500640869141
```

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: it seems not bad

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:408]: do we really need it? the constructor of SelectSQL already checks it?

**Review by sfc-gh-jdu — APPROVED** [review]

**Comment by sfc-gh-jdu** [comment]
Let's get Yijun's approval before merging

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:408]: Without this, select statements create a new `SelectSQL` and assign column_states = self.column_states.

In case when `self._sql_query` is `SELECT SYSTEM$WAIT(3)`, processing `column_states` triggers a describe query and it ends up in a blocking call. This was causing async test jobs failing.

After writing this explanation, I feel like we should only set `new.column_states = self.column_states` if `self.column_states is not None`

File: src/snowflake/snowpark/_internal/telemetry.py
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/_internal/telemetry.py:159]: How about testing 500, 1000 and 5000 logical plans?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:408]: yeah make sense

**Comment by sfc-gh-aalam** [comment]
Client performance when running union api when 50 columns and varying n calls:
Screenshot 2024-04-09 at 10 40 40 PM

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**
